### PR TITLE
varlink: make API public as "sd-varlink.h"

### PR DESCRIPTION
**Pull request by poettering**
After sd-json is already public now, let's also make sd-varlink public as part of libsystemd.

### CODE REVIEW

File: src/libsystemd/sd-json/sd-json.c [hunk 1]
```diff
@@ -1477,7 +1477,8 @@ _public_ int sd_json_variant_equal(sd_json_variant *a, sd_json_variant *b) {
 }
 
 _public_ void sd_json_variant_sensitive(sd_json_variant *v) {
-        assert(v);
+        if (!v)
+                return;
```
Comment (YHNdnzj) [inline src/libsystemd/sd-json/sd-json.c:1481]: I think we should use assert_return here, since this is a setter.

Comment (poettering) [reply src/libsystemd/sd-json/sd-json.c:1481]: for public apis we are usually relatively friendly. And conceptually, turning on sensitivity of a NULL object is kinda redundant. If you have no data, no data can be linked. Hence I figured we should be nice to users here and just make it a NOP.

File: src/libsystemd/sd-varlink/varlink-internal.h [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+}
+
+static inline const char *varlink_server_description(sd_varlink_server *s) {
+        return (s ? s->description : NULL) ?: "varlink";
+}
+
+VarlinkServerSocket *varlink_server_socket_free(VarlinkServerSocket *ss);
+DEFINE_TRIVIAL_CLEANUP_FUNC(VarlinkServerSocket *, varlink_server_socket_free);
+
+int varlink_acquire_pidfd(sd_varlink *v);
```
Comment (YHNdnzj) [inline src/libsystemd/sd-varlink/varlink-internal.h]: I think we should expose sd_varlink_get_peer_pidfd

File: src/systemd/_sd-common.h [hunk 3]
```diff
@@ -105,4 +105,11 @@ typedef void (*_sd_destroy_t)(void *userdata);
         _SD_##id##_INT64_MIN = INT64_MIN,    \
         _SD_##id##_INT64_MAX = INT64_MAX
 
+/* In GCC 14 (C23) we can force enums to have the right types, and not solely rely on language extensions anymore */
+#if ((__GNUC__ >= 14) || (__STDC_VERSION__ >= 202311L)) && !defined(__cplusplus)
+#define _SD_ENUM_TYPE_S64(id) id : int64_t
```
Comment (YHNdnzj) [inline src/systemd/_sd-common.h]:
```suggestion
#  define _SD_ENUM_TYPE_S64(id) id : int64_t
```

File: src/systemd/_sd-common.h [hunk 4]
```diff
@@ -105,4 +105,11 @@ typedef void (*_sd_destroy_t)(void *userdata);
...
+#else
+#define _SD_ENUM_TYPE_S64(id) id
```
Comment (YHNdnzj) [inline src/systemd/_sd-common.h]: also here

File: src/systemd/sd-varlink-idl.h [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+#define SD_VARLINK_DEFINE_OUTPUT_BY_TYPE(_name, _named_type, _field_flags) \
+        { .name = #_name, .field_type = SD_VARLINK_NAMED_TYPE, .named_type = #_named_type, .symbol = &vl_type_ ## _named_type, .field_flags = (_field_flags), .field_direction = SD_VARLINK_OUTPUT }
+
+#define SD_VARLINK_DEFINE_ENUM_VALUE(_name)                             \
+        { .name = #_name, .field_type = SD_VARLINK_ENUM_VALUE }
+
+#define SD_VARLINK_FIELD_COMMENT(text)                                  \
+        { .name = "" text, .field_type = _SD_VARLINK_FIELD_COMMENT }
+
+/* C++ older thatn C20+ does not support __VA_OPT__(), but we really need it here. */
```
Comment (YHNdnzj) [inline src/systemd/sd-varlink-idl.h]:
```suggestion
/* C++ older than C20+ does not support __VA_OPT__(), but we really need it here. */
```

File: src/shared/varlink-serialize.h [hunk 6]
```diff
@@ -0,0 +1,10 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+#pragma once
+
+#include <stdio.h>
+
+#include "fdset.h"
+#include "sd-varlink.h"
```
Comment (YHNdnzj) [inline src/shared/varlink-serialize.h]: Mixed type of includes?

File: src/libsystemd/sd-varlink/varlink-idl-util.h [hunk 7]
```diff
@@ -0,0 +1,24 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+#pragma once
+
+#include "sd-varlink-idl.h"
+#include "macro.h"
```
Comment (YHNdnzj) [inline src/libsystemd/sd-varlink/varlink-idl-util.h]: mixed type also here?

File: src/libsystemd/sd-varlink/varlink-idl-util.c [hunk 8]
```diff
@@ -0,0 +1,3 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+
+#include "varlink-idl-util.h"
```
Comment (YHNdnzj) [inline src/libsystemd/sd-varlink/varlink-idl-util.c]: Hmm, why is this file needed?

File: src/libsystemd/sd-varlink/varlink-internal.h [hunk 9]
```diff
@@ -0,0 +1,249 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+#pragma once
+
+#include <sys/socket.h>
+
+#include "sd-event.h"
+#include "sd-varlink.h"
+
+#include "list.h"
+#include "hashmap.h"
```
Comment (YHNdnzj) [inline src/libsystemd/sd-varlink/varlink-internal.h]: not sorted

File: src/libsystemd/sd-varlink/varlink-internal.h [hunk 10]
```diff
@@ -0,0 +1,249 @@
...
+
+static inline const char *varlink_description(sd_varlink *v) {
+        return (v ? v->description : NULL) ?: "varlink";
+}
+
+static inline const char *varlink_server_description(sd_varlink_server *s) {
+        return (s ? s->description : NULL) ?: "varlink";
+}
+
+VarlinkServerSocket *varlink_server_socket_free(VarlinkServerSocket *ss);
```
Comment (YHNdnzj) [inline src/libsystemd/sd-varlink/varlink-internal.h]:
```suggestion
VarlinkServerSocket* varlink_server_socket_free(VarlinkServerSocket *ss);
```

File: src/libsystemd/sd-varlink/sd-varlink.c [hunk 11]
```diff
@@ -3,10 +3,12 @@
 #include <malloc.h>
 #include <poll.h>
 
-#include <sd-daemon.h>
+#include "sd-daemon.h"
+#include "sd-varlink.h"
 
 #include "alloc-util.h"
 #include "errno-util.h"
+#include "errno-list.h"
```
Comment (YHNdnzj) [inline src/libsystemd/sd-varlink/sd-varlink.c:10]: not sorted

File: src/libsystemd/sd-varlink/varlink-util.c [hunk 12]
```diff
@@ -0,0 +1,136 @@
...
+        va_end(ap);
+        if (r < 0)
+                return r;
+
+        int ret = 1;
+        sd_varlink *link;
+        SET_FOREACH(link, s)
+                RET_GATHER(ret, sd_varlink_notify(link, parameters));
+
+        return ret;
```
Comment (YHNdnzj) [inline src/libsystemd/sd-varlink/varlink-util.c:111]: Hmm, just use `r` here?

File: src/systemd/sd-varlink.h [hunk 13]
```diff
@@ -0,0 +1,296 @@
...
+__extension__ typedef enum _SD_ENUM_TYPE_S64(sd_varlink_reply_flags_t) {
+        SD_VARLINK_REPLY_ERROR     = 1 << 0,
+        SD_VARLINK_REPLY_CONTINUES = 1 << 1,
+        SD_VARLINK_REPLY_LOCAL     = 1 << 2,
+        _SD_ENUM_FORCE_S64(SD_VARLINK_REPLY)
+} sd_varlink_reply_flags_t;
+
+__extension__ typedef enum _SD_ENUM_TYPE_S64(sd_varlink_method_flags_t) {
+        SD_VARLINK_METHOD_ONEWAY = 1 << 0,
+        SD_VARLINK_METHOD_MORE   = 2 << 1,
```
Comment (YHNdnzj) [inline src/systemd/sd-varlink.h]: Hmm?

File: src/libsystemd/sd-json/sd-json.c [hunk 14]
```diff
@@ -5133,7 +5137,7 @@ _public_ int sd_json_dispatch_variant(const char *name, sd_json_variant *variant
 
 _public_ int sd_json_dispatch_variant_noref(const char *name, sd_json_variant *variant, sd_json_dispatch_flags_t flags, void *userdata) {
         sd_json_variant **p = ASSERT_PTR(userdata);
-        assert(variant);
+        assert_return(variant, -EINVAL);
```
Comment (YHNdnzj) [inline src/libsystemd/sd-json/sd-json.c:5141]: insert a newline?

File: src/libsystemd/sd-varlink/sd-varlink-idl.c
Comment (YHNdnzj) [inline src/libsystemd/sd-varlink/sd-varlink-idl.c:1]: functions in this file are not marked as _public_? Are they intended to be used externally?

File: src/libsystemd/sd-varlink/varlink-idl-util.c (hunk 8 above)
Comment (poettering) [reply src/libsystemd/sd-varlink/varlink-idl-util.c]: it is not

File: src/libsystemd/sd-varlink/sd-varlink-idl.c
Comment (poettering) [reply src/libsystemd/sd-varlink/sd-varlink-idl.c:1]: nope, i just fucked up. three of the symbols should have `_public_`. will fix.

**Comment by poettering** [comment]
Force pushed new version, with all raised issues addressed. PTAL!

**Review by DaanDeMeyer — APPROVED** [review]

File: src/libsystemd/sd-varlink/varlink-util.c (hunk 12 above)
Comment (YHNdnzj) [reply src/libsystemd/sd-varlink/varlink-util.c:111]: Not addressed?

**Comment by bluca** [comment]
There are some functions which are now public APIs which lack test coverage, please add some tests for then, even if superficial:

https://coveralls.io/builds/68688207/source?filename=src%2Fshared%2Fvarlink.c

File: src/libsystemd/sd-varlink/varlink-util.c (hunk 12 above)
Comment (poettering) [reply src/libsystemd/sd-varlink/varlink-util.c:111]: https://github.com/systemd/systemd/pull/33753
